bin/arrayNormalization.R

```diff
@@ -0,0 +1,374 @@
+				dataSet <- try({read.maimages(files, source="agilent", green.only=TRUE)})
+			}
+			else if(mode == "agil2") {
+				
+				print("Reading Agilent 2-colour data")
+				dataSet <- try({read.maimages(files, source="agilent")})
+			}
+			# check it worked, if not, return the error.
+			if(class(dataSet) == "try-error") { return(dataSet) }
+			
+			# Check if there's a microRNA miRBase probeset list and if so,
+			# subset based on that list.
+			if(miRBaseFile != 0) { dataSet <- subsetProbes(dataSet, miRBaseFile) }
+			
+			# Background correction.
+			# Using method outlined in Section 6.1 of Limma Users Guide (25
+			# September 2012 version), to adjust the foreground adaptively for
+			# the background intensities, to produce strictly positive adjusted
+			# intensities, i.e. avoiding negative or zero values in the
+			# resulting adjusted intensities.
+			# NB: this might get masked by oligo package's backgroundCorrect
+			# method if they are both loaded (here they shouldn't be so don't
+			# call limma's explicitly).
+			print("Background correcting")
+			dataSet <- backgroundCorrect(dataSet, method="normexp", offset=50)
+			
+			if(mode == "agil1") {
+				
+				# Between-array normalization for 1-colour.
+				normData <- normalizeBetweenArrays(dataSet, method="quantile")
+			}
+
+			else if(mode == "agil2") {
+				# Within-array normalization for 2-colour.
+				# This gives the M-value and A-value for each gene on the array.
+				# M = log2(Cy5) - log2(Cy3)
+				# A = 0.5[log2(Cy5) + log2(Cy3)]
+				print("Running within-array normalization")
+				normData <- normalizeWithinArrays(dataSet, method="loess")
+			}
+
+			# average for probe duplicates -- some probes are on the array
+			# multiple times in different locations.
+			print("Averaging duplicated probes")
+			normData <- avereps(normData, ID=normData$genes$ProbeName)
+			
+			# Write results to file
+			if(mode == "agil1") {
+				# For single channel Agilent data, this is just the
+				# quantile-normalized expressions (analogous to RMA-normalized
+				# expressions from Affymetrix data).
+				
+				# Make a dataframe with probe IDs and normalized expressions.
+				normExprs <- data.frame(cbind(normData$genes$ProbeName, normData$E))
+				# Add assayNames as column headings
+				colnames(normExprs) <- c("DesignElementAccession", assayNames)
+
+				print("Writing normalized expressions matrix")
+				write.table(normExprs, file=outFile, sep="\t", quote=FALSE, row.names=FALSE)
+				
+				# Now files are written, function returns and script exits.
+				return("Agilent 1-colour normalization finished")
+			}
+			else if(mode == "agil2") {
+				# For two channel Agilent data, this is logFCs (M-values) and
+				# average intensities (A-values).
+
+				# data frames for M-values and A-values. Put rownames (design
+				# element IDs) as the first column and name it
+				# "DesignElementAccession".
+				mValues <- data.frame(normData$M)
```

Comment:
Again here, I would avoid the cbind pattern and create a named data.frame

Reply:
I would suggest in general as well to prefer data.table over data.frame.

Reply:
assayNames is an object holding assay names. Not a column name.